### ISSUE TITLE
Admin App - Conflict resolution

### ISSUE THREAD

**Issue by BincyJ** [unfoldingWord/tc-create-app#518]
As an admin, I should be able to view the conflicts between latest master repo and a user branch.

**Comment by mandolyte** [comment]
# Issue 518 "SPIKE: Admin App - Conflict resolution"

- Issue: https://github.com/unfoldingWord/tc-create-app/issues/518
- Drive location for diagram: https://drive.google.com/file/d/1HQqLWeaTj6aLCzJrIGQCAfA-zgwcpjmh/view?usp=sharing_eip&ts=5fb3d6ec
- Workflow diagram: https://app.diagrams.net/#G1HQqLWeaTj6aLCzJrIGQCAfA-zgwcpjmh
- Rich's PoC work: https://unfoldingword.zulipchat.com/#narrow/stream/209457-SOFTWARE/topic/DCS.20POC.20-.20Merge.20branches.20with.20API/near/215030799
- The PoC artifacts at https://unfoldingword-dev/dcs-api-merge-poc 

# Essentials

The term "merge" is used in two senses in this document. 
- Branches can be merged
- Changes to a single file can be merged. The word "resolve" is used to make this clear when needed.

In most cases, the context will be clear about which is being discussed.

1. User needs to know if a file in their branch has changed on master
2. User needs to be able to merge master into their branch when preceding is true and resolve any conflicts
3. User needs to be able to "submit a PR" for their changes.

# Flow of changes into the Master branch

## At start

- When a user begins making changes, a branch is created from the master branch.
- Later, when the user wishes to save the changes, they will click the "Save" button.
- This will save the changes to the user's branch.
- Later there will be an attempt to merge the user's branch to the master branch.

NOTE 1: if the master has not been changed subsequent to the user's branch creation, 
then the user's branch will always be mergeable. (branch merged)

NOTE 2: if the master branch has been changed, but the change does not involve any file they 
have edited in their branch, then the user's branch will always be mergeable. (branch merged)

The remainder of this describes what happens when the above two NOTES are not the case, namely, 
the master version of a file has changed and the user has also made changes to the same file.
In this case, the two sets of changes must be resolved. This may or may not involve conflicts
(i.e., changes made to the same lines).

The following describes the process for a single file, but it must be done for all files 
where the changes must be merged (file merge).

*Gitea APIs needed:*
Detect whether master and branch are mergeable without conflict. If there are conflicts, then
by definition, some subset of the files changed will have to be resolved.

## Case 1: No conflicts

This will be the case when changes are not to the same or not to adjacent lines. 
If this is the case, then the two versions of the file are considered "mergeable".

*Gitea APIs needed:*
Merge master into branch.

In this case, the user can update their files with the newer files from master and may
continue making changes with the latest changes made by others.

## Case 2: Conflicts

If Gitea cannot automatically merge due to conflicts (same or adjacent lines have been 
changed in both versions of the file), then the user must be able to resolve the 
conflicts. There are three possible actions:

1. The user selects a change from master version to replace their own
2. The user retains their change, discarding the version of the change from master
3. The user opts to combine in some fashion the two changes

In the third case, we expect the user to abandon the merge, make the synthesized change
in their branch, then attempt the merge again. This process essentially reduces the
possible actions to two in the end.

*Gitea APIs needed:*
A comparison difference result which can be used in an application/RCL to guide the user
thru resolving the conflicts.

## At end

Once the user is done, then the changes must be submitted as a Pull Request (PR).
If the user continues making changes in their branch, then the PR will automatically
include new changes. This is an important training issue.

*Gitea APIs needed:*
A way to create the PR for the user's branch.

**Comment by mandolyte** [comment]
Found an interesting website that converts curl commands into Javascript fetch():

This website converts curl commands into Javascript fetch:
https://kigiri.github.io/fetch/

For example, this:
```
curl POST "https://github.com/api/v1/repos/$org/$repo/pulls?access_token=$token" -H "accept: application/json" -H "Content-Type: application/json" -d '{ "base": "master", "body": "Merging user branch into master", "head": "$branch", "title": "$branch into master"}'
```

becomes this:
```
fetch("https://github.com/api/v1/repos/$org/$repo/pulls?access_token=$token", {
  body: "{ \"base\": \"master\", \"body\": \"Merging user branch into master\", \"head\": \"$branch\", \"title\": \"$branch into master\"}",
  headers: {
    Accept: "application/json",
    "Content-Type": "application/json"
  },
  method: "POST"
})
```

**Comment by mandolyte** [comment]
# An alternate view of the translation workflow (partial)

## Overview

In this alternate view, the translate process at a high level could be:

The translator:
- selects a file to translate or to update
- makes changes to the file
- save changes to the file
- this may continue for for some time
- proposes to replace the orginal with their file
- if discovered that someone else is also working with the same file, then the first translator's changes are merged into the second translator's changes, reconciling any conflicts that may arise.

The reviewer:
- collaborates with translator to resolve any questions
- approves the changes, which...
- replaces the orginal with the approved (updated) version

This alternate view is, in essence, a document life-cycle. This is not how Gitea works, but it is possible to layer this life-cycle over the top of Gitea.

## Solution Description

1. Let each branch be associated to the translator and to the file being edited. That is, the name of the branch must be a combination of user and file name. So if the user is working on three files, then that will involve three branches.
2. If two users are working on same file, for example, different chapters in the same book, then this results in two branches being created.
3. There is a reviewer role who will be responsible for approving changes and updating the master branch. 

See scenarios below for details.

## Scenarios

In the following, unless otherwise noted, the book of Ruth is used as book being created or updated. Alice and Bob are the translators and Charles is the reviewer. The acronym PR stands for Pull Request (a Git term). 

If the resource being worked on is a book-oriented resource, then the branch created is a combination of the bookd id and the user's id. For example, "rut-alice".

If the resource is a tA or tN, then the full name of the article is used. For example, "figs-metaphor-bob".

### Initial Creation of a Translated File

_Description_ In this case there are no conflicts to resolve with other translators and no actionable feedback from the reviewer.

_Steps_ In this case, the translator uses tc-create to create a file; saves the work; submits a PR. The reviewer merges the changes into the master branch.

_Details_ Alice selects Ruth to translate. A branch named "rut-alice" is created. Updates are made and saved. After some time passes, the translation is ready for review. Alice submits a PR. The reviewer is assigned along with any comments by Alice about the PR. The reviewer, having no feedback, merges Alice's branch into the master branch. Since this book is new and it is the only content in Alice's branch, it simply creates the file in the master branch. The working branch "rut-alice" is deleted.

### Revision to a Translated File

_Description_ This describes the case where there are no conflicts to resolve with other translators and no actionable feedback from the reviewer.

_Steps_ In this case, the translator uses tc-create to open and update a file; saves the work; submits a PR. The reviewer merges the changes into the master branch.

_Details_ Alice selects Ruth to update. A branch named "rut-alice" is created. Updates are made and saved. After some time passes, the updated translation is ready for review. Alice submits a PR. The reviewer is assigned along with any comments by Alice about the PR. The reviewer, having no feedback, merges Alice's branch into the master branch. Since this book exists in the master branch and this book is the sole content in the Alice's working branch, the effect is simply to replace the file in the master branch.

### Two translators working to Translate the same File, but no conflicts

_Description_ This describes the case where there are two translators are working on the same file, but are working in different chapters with no conflicts; and with no actionable feedback from the reviewer.

_Steps_ In this case, the translators uses tc-create to open and update the same file and save their work. At some point, one of them will attempt to submit a PR. At this point, it is discovered that that someone else is also working in the same file; this causes the PR to fail to be created. At this point, the first user's changes are merged into the still on-going changes of the second user. Since there are no conflicts, then this will be fairly straighforward since both users' changes will be accepted into the second user's branch. At this point, the second user's branch will incorporate all the first user's changes.

Later the other translator will attempt to submit a PR. Since master has not changed since they began, the PR will succeed and the changes will replace the file in the master branch.

_Details_ Alice selects Ruth to translate. A branch named "rut-alice" is created. Bob does the same, creating a branch named "rut-bob". Updates are made and saved by both Alice and Bob. After some time passes, Bob attempts to submits a PR. However, it is discovered that Alice is also making changes to the same file; the PR fails to be created. Then Bob's changes are merged into Alice's changes, so that Alice's file now has both hers and Bob's changes. Bob's branch is deleted. She continues to work and submits a PR when finished.

Since the file in the master branch is unchanged since Alice started, the PR succeeds and the new content replaces the old in the master branch. Alice's branch is deleted.

### Two translators working to Translate the same File, with conflicts

_Description_ This path describes the case where there are two translators are working on the same file and end up translating the same portion of the file differently; and with no actionable feedback from the reviewer.

_Steps_ In this case, the translators uses tc-create to open and update the same file; save their work. At some point, one of them will attempt to submit a PR. At this point, it is discovered that that someone else is also working in the same file. At this point, the first user's changes are merged into the still on-going changes of the second user. Since there are  conflicts, then this could be fairly complex, especially if the desired result is a combination of both translators' changes. 

Once the second user's branch has the desired reconciled content, then the first user's branch is deleted.

Later the other translator will attempt to submit a PR. Since master has not changed since they began, the PR will succeed and the changes will replace the file in the master branch.

_Details_ Alice selects Ruth to translate. A branch named "rut-alice" is created. Bob does the same, creating a branch named "rut-bob". Updates are made and saved by both Alice and Bob. In particular, both have made changes to the same part of the the resource file. After some time passes, Bob attempts to submits a PR. However, it is discovered that Alice is also making changes to the same file. So Bob's changes are merged into Alice's changes, reconciling the conflicts so that Alice's file now has both hers and Bob's changes, with any conflicts resolved. Bob's branch is deleted. Alice continues to work and submits a PR when finished.

Since the file in the master branch is unchanged since Alice started, the PR succeeds and the new content replaces the old in the master branch. Alice's branch is deleted.

### Reviewer has changes

If, at any time after PR is submitted, the reviewer wants changes made, then the reviewer collaborates with the translator to make the desired changes in their branch. The content associated with the PR will update as well. This continues until the reviewer and translator are satisfied. At this point, the PR is used to merge the user's branch into the master branch.

**Comment by mandolyte** [comment]
@BincyJ @birchamp Because it seems that our users have a hard time dealing with Git, especially in how branches work, I created the above which hides those aspects and presents a book or article view of the translation tasks. But this would require some changes to tc-create that go beyond the scope of this spike. I might add it would make the code for resolving conflicts easier as well.

However, if there is no interest in tackling how tc-create uses Git to manage the translation workflow, then for this spike I need to deal with the complexities as they are.  So while I think there is value in looking at things like a commercial document management system, I suspect that is not the highest priority at this time.

So I'll pursue how to work with tc-create/Git as-is for now. I hope to have the general description of what it needed mostly completed by tomorrow. 

Is there a "definition of done" for this spike? Thanks!

**Comment by mandolyte** [comment]
> As an admin, I should be able to view the conflicts between latest master repo and a user branch.
> 
> https://docs.google.com/document/d/10pXKcA6k7GWqD93lNZ_T_xhD2jmIamgpdv1foXJvW44/edit

This document needs to be revised based on user preference for a reviewer...

**Comment by birchamp** [comment]
@birchamp we need to decide if user-chapter branches are the correct solution to DCS workflow.
